File: matlab/src/api/pfqn/pfqn_bsfcfs.m
%{
%{
 % @file pfqn_bsfcfs.m
 % @brief Bard-Schweitzer approximate MVA for FCFS scheduling with weighted priorities.
%}
%}

%{
%{
 % @brief Bard-Schweitzer approximate MVA for FCFS scheduling with weighted priorities.
 % @fn pfqn_bsfcfs(L, N, Z, tol, maxiter, QN, weight)
 % @param L Service demand matrix.
 % @param N Population vector.
 % @param Z Think time vector (default: zeros).
 % @param tol Convergence tolerance (default: 1e-6).
 % @param maxiter Maximum number of iterations (default: 1000).
 % @param QN Initial queue length matrix (default: uniform distribution).
 % @param weight Weight matrix for relative priorities (default: ones).
 % @return XN System throughput.
 % @return QN Mean queue lengths.
 % @return UN Utilization.
 % @return RN Residence times.
 % @return it Number of iterations performed.
%}
%}
function [XN,QN,UN,RN,it]=pfqn_bsfcfs(L,N,Z,tol,maxiter,QN,weight)
% [XN,QN,UN,RN]=PFQN_BSFCFS(L,N,Z,TOL,MAXITER,QN,WEIGHT)

if nargin<3%~exist('Z','var')
    Z=0*N;
end
if nargin<4%~exist('tol','var')
    tol = 1e-6;
end
if nargin<5%~exist('maxiter','var')
    maxiter = 1000;
end
[M,R]=size(L);
CN=zeros(M,R);
if nargin<6 %~exist('QN','var')
    QN=repmat(N,M,1)/M;
else
    QN = QN+eps; % 0 gives problems
end
if nargin<7 %~exist('weight','var')
    weight = ones(M,R);
end
XN=zeros(1,R);
UN=zeros(M,R);
relprio=zeros(M,R);
for it=1:maxiter
    QN_1 = QN;
    for ist=1:M
        for r=1:R
            relprio(ist,r) = (QN(ist,r)*weight(ist,r));
        end
    end
    for r=1:R
        for ist=1:M
            CN(ist,r) = L(ist,r);
            for s=1:R
                if s~=r
                    % FCFS approximation
                    CN(ist,r) = CN(ist,r) + L(ist,s)*QN(ist,s)*relprio(ist,s)/relprio(ist,r);
                else
                    CN(ist,r) = CN(ist,r) + L(ist,r)*QN(ist,r)*(N(r)-1)/N(r)*relprio(ist,s)/relprio(ist,r);
                end
            end
        end
        XN(r) = N(r)/(Z(r)+sum(CN(:,r)));
    end
    for r=1:R
        for ist=1:M
            QN(ist,r) = XN(r)*CN(ist,r);
        end
    end
    for r=1:R
        for ist=1:M
            UN(ist,r) = XN(r)*L(ist,r);
        end
    end
    if max(abs(1-QN./QN_1)) < tol
        break
    end
end
RN = QN ./ repmat(XN,M,1);
end

File: jar/src/main/java/jline/api/pfqn/mva/Pfqn_bsfcfs.java
/**
 * @file Bard-Schweitzer approximate MVA for FCFS scheduling with weighted priorities
 *
 * @since LINE 3.0
 */
package jline.api.pfqn.mva;

import org.apache.commons.math3.util.FastMath;

import jline.io.Ret;
import jline.util.Maths;
import jline.util.matrix.Matrix;

public final class Pfqn_bsfcfs {
    private Pfqn_bsfcfs() {}

    /**
     * Bard-Schweitzer approximate MVA for FCFS scheduling with weighted priorities.
     */
    public static Ret.pfqnAMVA pfqn_bsfcfs(Matrix L,
                                           Matrix N,
                                           Matrix Z,
                                           double tol,
                                           int maxiter,
                                           Matrix QN0,
                                           Matrix weight) {
        int M = L.getNumRows();
        int R = L.getNumCols();

        Matrix Zmat = (Z != null) ? Z : new Matrix(1, R);

        Matrix CN = new Matrix(M, R);
        Matrix QN;
        if (QN0 == null || QN0.isEmpty()) {
            QN = N.repmat(M, 1);
            for (int i = 0; i < QN.getNumRows(); i++) {
                for (int j = 0; j < QN.getNumCols(); j++) {
                    QN.set(i, j, QN.get(i, j) / M);
                }
            }
        } else {
            QN = QN0.copy();
            for (int i = 0; i < QN.getNumRows(); i++) {
                for (int j = 0; j < QN.getNumCols(); j++) {
                    QN.set(i, j, QN.get(i, j) + FastMath.ulp(1.0));
                }
            }
        }

        Matrix weightMatrix = (weight != null) ? weight : Matrix.ones(M, R);

        Matrix XN = new Matrix(1, R);
        Matrix UN = new Matrix(M, R);
        Matrix relprio = new Matrix(M, R);

        int it = 1;
        while (it <= maxiter) {
            Matrix QN_1 = new Matrix(QN);

            for (int ist = 0; ist < M; ist++) {
                for (int r = 0; r < R; r++) {
                    relprio.set(ist, r, QN.get(ist, r) * weightMatrix.get(ist, r));
                }
            }

            for (int r = 0; r < R; r++) {
                for (int ist = 0; ist < M; ist++) {
                    CN.set(ist, r, L.get(ist, r));
                    for (int s = 0; s < R; s++) {
                        if (s != r) {
                            CN.set(ist, r,
                                CN.get(ist, r) + L.get(ist, s) * QN.get(ist, s) * relprio.get(ist, s) / relprio.get(ist, r));
                        } else {
                            CN.set(ist, r,
                                CN.get(ist, r) + L.get(ist, r) * QN.get(ist, r) * (N.get(r) - 1) / N.get(r) * relprio.get(ist, s) / relprio.get(ist, r));
                        }
                    }
                }
                XN.set(r, N.get(r) / (Zmat.get(r) + Matrix.extractColumn(CN, r, null).elementSum()));
            }
            for (int r = 0; r < R; r++) {
                for (int ist = 0; ist < M; ist++) {
                    QN.set(ist, r, XN.get(r) * CN.get(ist, r));
                }
            }
            for (int r = 0; r < R; r++) {
                for (int ist = 0; ist < M; ist++) {
                    UN.set(ist, r, XN.get(r) * L.get(ist, r));
                }
            }

            double maxabs = Double.MIN_VALUE;
            for (int i = 0; i < QN.getNumRows(); i++) {
                for (int j = 0; j < QN.getNumCols(); j++) {
                    double absValue = FastMath.abs(1 - QN.get(i, j) / QN_1.get(i, j));
                    maxabs = Maths.max(maxabs, absValue);
                }
            }
            if (maxabs < tol) {
                break;
            }
            it++;
        }

        Matrix RN = XN.repmat(M, 1);
        for (int i = 0; i < RN.getNumRows(); i++) {
            for (int j = 0; j < RN.getNumCols(); j++) {
                RN.set(i, j, QN.get(i, j) / RN.get(i, j));
            }
        }
        return new Ret.pfqnAMVA(QN, UN, RN, null, CN, XN, it);
    }

    public static Ret.pfqnAMVA pfqn_bsfcfs(Matrix L, Matrix N, Matrix Z, double tol, int maxiter, Matrix QN0) {
        return pfqn_bsfcfs(L, N, Z, tol, maxiter, QN0, null);
    }

    public static Ret.pfqnAMVA pfqn_bsfcfs(Matrix L, Matrix N, Matrix Z, double tol, int maxiter) {
        return pfqn_bsfcfs(L, N, Z, tol, maxiter, null, null);
    }

    public static Ret.pfqnAMVA pfqn_bsfcfs(Matrix L, Matrix N, Matrix Z, double tol) {
        return pfqn_bsfcfs(L, N, Z, tol, 1000, null, null);
    }

    public static Ret.pfqnAMVA pfqn_bsfcfs(Matrix L, Matrix N, Matrix Z) {
        return pfqn_bsfcfs(L, N, Z, 1.0e-6, 1000, null, null);
    }

    public static Ret.pfqnAMVA pfqn_bsfcfs(Matrix L, Matrix N) {
        return pfqn_bsfcfs(L, N, null, 1.0e-6, 1000, null, null);
    }
}
